ambry-account/src/main/java/com/github/ambry/account/HelixAccountService.java

```diff
@@ -0,0 +1,492 @@
+   * </p>
+   */
+  @Override
+  public boolean updateAccounts(Collection<Account> accounts) {
+    checkOpen();
+    if (accounts == null) {
+      throw new IllegalArgumentException("accounts cannot be null");
+    }
+    if (hasDuplicateAccountIdOrName(accounts)) {
+      logger.debug("Duplicate account id or name exist in the accounts to update");
+      accountServiceMetrics.updateAccountErrorCount.inc();
+      return false;
+    }
+    long startTimeMs = System.currentTimeMillis();
+    logger.trace("Start updating to HelixPropertyStore with accounts={}", accounts);
+    boolean hasSucceeded = false;
+    // make a pre check for conflict between the accounts to update and the accounts in the local cache. Will fail this
+    // update operation for all the accounts if any conflict exists. There is a slight chance that the account to update
+    // conflicts with the accounts in the local cache, but does not conflict with those in the helixPropertyStore. This
+    // will happen if some accounts are updated but the local cache is not refreshed.
+    if (hasConflictWithCache(accounts)) {
+      logger.debug("Accounts={} conflict with the accounts in local cache. Cancel the update operation.", accounts);
+      accountServiceMetrics.updateAccountErrorCount.inc();
+    } else {
+      hasSucceeded = helixStore.update(FULL_ACCOUNT_METADATA_PATH, currentData -> {
+        ZNRecord newRecord;
+        if (currentData == null) {
+          logger.debug(
+              "ZNRecord does not exist on path={} in HelixPropertyStore when updating accounts. Creating a new ZNRecord.",
+              FULL_ACCOUNT_METADATA_PATH);
+          newRecord = new ZNRecord(String.valueOf(ZN_RECORD_ID));
+        } else {
+          newRecord = currentData;
+        }
+        Map<String, String> accountMap = newRecord.getMapField(ACCOUNT_METADATA_MAP_KEY);
+        if (accountMap == null) {
+          logger.debug("AccountMap does not exist in ZNRecord when updating accounts. Creating a new accountMap");
+          accountMap = new HashMap<>();
+        }
+        AccountInfoMap remoteAccountInfoMap;
+        try {
+          remoteAccountInfoMap = new AccountInfoMap(accountMap);
+        } catch (JSONException e) {
+          // Do not depend on Helix to log, so log the error message here.
+          logger.error("Exception occurred when building AccountInfoMap from accountMap={}", accountMap, e);
+          accountServiceMetrics.remoteDataCorruptionErrorCount.inc();
+          throw new IllegalStateException("Exception occurred when building AccountInfoMap from accountMap", e);
+        }
+        // if there is any conflict with the existing record, fail the update. Exception thrown in this updater will
+        // be caught by Helix and helixStore#update will return false.
+        if (hasConflictingAccount(accounts, remoteAccountInfoMap)) {
+          String message = "Updating accounts failed because one account to update conflicts with existing accounts";
```

Comment:
It is actually logging at two levels, but I agree that the upper level may not be necessary. Have removed this logging.